rfcs/rfc0010.md

```diff
@@ -0,0 +1,157 @@
+# A built-in for getting key-and-value pairs
+
+## Preamble
+
+    Author:  Ricardo Signes <rjbs@semiotic.systems>
+    Sponsor:
+    ID:      0010
+    Status:  Draft
+
+## Abstract
+
+This RFC proposes `kv`, a new function for returning the keys and values of a
+container.  Among other things, this allows key/value iteration on arrays as
+easily on hashes.
+
+## Motivation
+
+Right now, the `keys` and `values` functions operate on both arrays and hashes,
+giving the same semantics, if we accept that array indexes are effectively the
+keys of the array.  Pair slices, similarly, produce key/value pairs of array
+entries.  The sole missing component is that the behavior of a hash in list
+context is not easily reproduced.
+
+Providing this doesn't just fulfill a sense of completionism, but also has an
+immediate practical benefit.  With v5.36.0 poised to add n-at-a-time foreach,
+easily getting a list of index/value pairs from an array makes iteration over
+them easy.
+
+## Rationale
+
+If we start with the specific case of iterating over the indexes and values of
+an array using two-target foreach, we might write this:
+
+```perl
+for my ($i, $value) (%array[ keys @array ]) {
+  say "$i == $value";
+}
+```
+
+This is tolerable, but a bit verbose.  If we bury our target array deep in a
+structure, we get this:
+
+```perl
+for my ($i, $value) ($alpha->{beta}->[0]->%[ keys $alpha->{beta}->[0]->@* ]) {
+  say "$i == $value";
+}
+```
+
+This is pretty bad.
+
+With `kv`, we write this:
+
+```perl
+for my ($i, $value) (kv $alpha->{beta}->[0]->@*) {
+  say "$i == $value";
+}
+```
+
+This is probably about as simple as this can get without some significant new
+addition to the language.
+
+## Specification
+
+Like `keys` or `values`, `kv` can be called with a hash expression or array
+expression.  Other types of arguments are compile time errors.
+
+In scalar context, `kv` returns the number of entries in its argument, just
+like `keys` or `values`.
+
+In list context, `kv %h` is identical in behavior to `%h` in list context.
+
+In list context, `kv @arr` returns a list twice the size of the array, zipping
```

Comment:
_zipping_ is a wholly different operation. What I recommend here, is to speak of `kv` as return a _tuple_ of size 2. I.e., give a `%h` or `@a`, one should expect to get each _pair_ :

* (0,1)
* (2,3)
* ...
* (`$#a-1`, `$#a`)

Reply:
I wanted to mention after some thought, `kv` _can_ be thought of as an _unzip_ operation. That said, a true _zip_ of 2 lists would be a useful thing. See also, https://www.geeksforgeeks.org/zip-in-python/